File: Scripts/board.py
from boards import white_pieces, black_pieces, pawns, knights, bishops, rooks, queens, kings


def print_board(board):
    print("  ---------------------------------------")
    for i in range(8):
        row = board[(i * 8):(i * 8) + 8][::-1]
        for j in range(8):
            print(" | " + row[j], end="")
        print(" | ")
        print("  ---------------------------------------")


def compute_board():
    board = ["  " for square in range(64)]
    white_pawns = pawns & white_pieces
    black_pawns = pawns & black_pieces
    white_knights = knights & white_pieces
    black_knights = knights & black_pieces
    white_bishops = bishops & white_pieces
    black_bishops = bishops & black_pieces
    white_rooks = rooks & white_pieces
    black_rooks = rooks & black_pieces
    white_queens = queens & white_pieces
    black_queens = queens & black_pieces
    white_king = kings & white_pieces
    black_king = kings & black_pieces

    for i in range(64):
        if ("0" * (64 - len(bin(white_pawns)[2:])) + bin(white_pawns)[2:])[i] == "1":
            board[i] = "wp"

        if ("0" * (64 - len(bin(black_pawns)[2:])) + bin(black_pawns)[2:])[i] == "1":
            board[i] = "bp"

        if ("0" * (64 - len(bin(white_knights)[2:])) + bin(white_knights)[2:])[i] == "1":
            board[i] = "wn"

        if ("0" * (64 - len(bin(black_knights)[2:])) + bin(black_knights)[2:])[i] == "1":
            board[i] = "bn"

        if ("0" * (64 - len(bin(white_bishops)[2:])) + bin(white_bishops)[2:])[i] == "1":
            board[i] = "wb"

        if ("0" * (64 - len(bin(black_bishops)[2:])) + bin(black_bishops)[2:])[i] == "1":
            board[i] = "bb"

        if ("0" * (64 - len(bin(white_rooks)[2:])) + bin(white_rooks)[2:])[i] == "1":
            board[i] = "wr"

        if ("0" * (64 - len(bin(black_rooks)[2:])) + bin(black_rooks)[2:])[i] == "1":
            board[i] = "br"

        if ("0" * (64 - len(bin(white_queens)[2:])) + bin(white_queens)[2:])[i] == "1":
            board[i] = "wq"

        if ("0" * (64 - len(bin(black_queens)[2:])) + bin(black_queens)[2:])[i] == "1":
            board[i] = "bq"

        if ("0" * (64 - len(bin(white_king)[2:])) + bin(white_king)[2:])[i] == "1":
            board[i] = "wk"

        if ("0" * (64 - len(bin(black_king)[2:])) + bin(black_king)[2:])[i] == "1":
            board[i] = "bk"

    return board


def precompute_all_moves(piece):
    moves = []
    if piece == "white_pawn":
        white_pawn_offsets = [8, 16]
        for current_square in range(64):
            if 7 < current_square < 16:
                moves.append(((1 << current_square) << white_pawn_offsets[0]) + ((1 << current_square) << white_pawn_offsets[1]))
            elif current_square < 56:
                moves.append(((1 << current_square) << white_pawn_offsets[0]))
            else:
                moves.append(0)

    if piece == "black_pawn":
        black_pawn_offsets = [-8, -16]
        for current_square in range(64):
            if 47 < current_square < 56:
                moves.append((1 << (current_square + black_pawn_offsets[0])) + (1 << (current_square + black_pawn_offsets[1])))
            elif current_square > 7:
                moves.append(1 << (current_square + black_pawn_offsets[0]))
            else:
                moves.append(0)

    if piece == "knight":
        knight_offsets = [-17, -15, -10, -6, 6, 10, 15, 17]
        for current_square in range(64):
            current_move = 0
            for offset in knight_offsets:
                if 63 >= (current_square + offset) >= 0:
                    if abs((current_square % 8) - (current_square + offset) % 8) <= 2 and abs(current_square // 8 - (current_square + offset) // 8) <= 2:
                        current_move |= 1 << (current_square + offset)

            moves.append(current_move)

    if piece == "bishop":
        bishop_offsets = []
        for offset in range(1, 8):
            bishop_offsets.append(offset * 7)
            bishop_offsets.append(-offset * 7)

            bishop_offsets.append(offset * 9)
            bishop_offsets.append(-offset * 9)
        for current_square in range(64):
            current_move = 0
            for offset in bishop_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == vertical_difference:
                        current_move |= 1 << (current_square + offset)

            moves.append(current_move)

    if piece == "rook":
        rook_offsets = []
        for offset in range(1, 8):
            rook_offsets.append(offset * 8)
            rook_offsets.append(-offset * 8)

            rook_offsets.append(offset)
            rook_offsets.append(-offset)

        for current_square in range(64):
            current_move = 0
            for offset in rook_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == vertical_difference:
                        current_move |= 1 << (current_square + offset)

            moves.append(current_move)

    if piece == "queen":
        queen_diagonal_offsets = []
        queen_straight_offsets = []
        for offset in range(1, 8):
            queen_straight_offsets.append(offset)
            queen_straight_offsets.append(-offset)

            queen_straight_offsets.append(offset * 8)
            queen_straight_offsets.append(-offset * 8)

            queen_diagonal_offsets.append(offset*7)
            queen_diagonal_offsets.append(-offset*7)

            queen_diagonal_offsets.append(offset * 9)
            queen_diagonal_offsets.append(-offset * 9)

        for current_square in range(64):
            current_move = 0

            for offset in queen_straight_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == 0 or vertical_difference == 0:
                        current_move |= 1 << (current_square + offset)

            for offset in queen_diagonal_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == vertical_difference:
                        current_move |= 1 << (current_square + offset)

            moves.append(current_move)

    if piece == "king":
        king_diagonal_offsets = [-9, -7, 7, 9]
        king_straight_offsets = [-8, -1, 1, 8]

        for current_square in range(64):
            current_move = 0

            for offset in king_straight_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == 0 or vertical_difference == 0:
                        current_move |= 1 << (current_square + offset)

            for offset in king_diagonal_offsets:
                if 63 >= (current_square + offset) >= 0:
                    horizontal_difference = abs((current_square % 8) - ((current_square + offset) % 8))
                    vertical_difference = abs((current_square // 8) - ((current_square + offset) // 8))
                    if horizontal_difference == vertical_difference:
                        current_move |= 1 << (current_square + offset)

            moves.append(current_move)

    return moves


# pieces
'''
white pawn = 0b1000
black pawn = 0b0000

knight = 0b1001

bishop = 0b1010

rook = 0b1011

queen = 0b1100

king = 0b1101
'''

File: Scripts/src/board.rs
// A number representation of all pieces and colors
pub const EMPTY: u8 = 0;
pub const PAWN: u8 = 1;
pub const ROOK: u8 = 2;
pub const KNIGHT: u8 = 3;
pub const BISHOP: u8 = 4;
pub const QUEEN: u8 = 5;
pub const KING: u8 = 6;

pub const WHITE: u8 = 8;
pub const BLACK: u8 = 16;
// A collection of bitmaps to represent the board in a simple and efficient manner
#[derive(PartialEq)]
pub struct Board {
    pub pawns: u64,
    pub rooks: u64,
    pub knights: u64,
    pub bishops: u64,
    pub queens: u64,
    pub kings: u64,

    pub white_pieces: u64,
    pub black_pieces: u64,
    // Number representing a square where en passant is possible, most likely 0
    // 0 does not encounter any problems because there is no situation where en passant on square 0 is a possibility
    pub en_passant: i8,
    // Four bit number where each bit represents if one specific castle is legal
    // First bit is kings side castling with white
    // Second bit is queens side castling with white
    // Third bit is kings side castling with black
    // Fourth bit is queens side castling with black
    pub castling: u8,
    // 8 == white, 16 == black
    pub turn: u8,
    // Square of promotion, default = -1
    pub promoted: i8,
    pub promoted_piece: u8,
    pub current_move: u64
}
// A way of representing a piece
pub struct Piece {
    pub color: u8,
    pub piece_type: u8,
}
// Represents a square as two numbers, the color and piece on it
// If no piece is present piece_type and color will be EMPTY which is equal to 0
pub fn get_piece(board: &Box<Board>, pos: i8) -> Piece {
    let color = if board.white_pieces & (1 << pos) > 0 { WHITE } else if board.black_pieces & (1 << pos) > 0 { BLACK } else { EMPTY };
    let mut piece_type: u8 = 0;
    if color == EMPTY {
        piece_type = EMPTY;
    } else if (1 << pos) & board.pawns > 0 {
        piece_type = PAWN;
    } else if (1 << pos) & board.rooks > 0 {
        piece_type = ROOK;
    } else if (1 << pos) & board.knights > 0 {
        piece_type = KNIGHT;
    } else if (1 << pos) & board.bishops > 0 {
        piece_type = BISHOP;
    } else if (1 << pos) & board.queens > 0 {
        piece_type = QUEEN;
    } else if (1 << pos) & board.kings > 0 {
        piece_type = KING;
    }
    Piece{color, piece_type}
}

pub fn get_king_pos(board: &Box<Board>, color: u8) -> i8 {
    let king_bit_board = if color == WHITE {
        board.white_pieces & board.kings
    } else {
        board.black_pieces & board.kings
    };
    for square in 0..64 {
        if king_bit_board & (1 << square) > 1 {
            return square;
        }
    }
    return -1
}

// Creates the starting positions bitboard
pub fn create_board() -> Box<Board> {
    let pawns: u64 = (((1 << 8) - 1) << 8) + (((1 << 8) - 1) << 48);
    let rooks: u64 = (1 << 0) + (1 << 7) + (1 << 56) + (1 << 63);
    let knights: u64 = (1 << 1) + (1 << 6) + (1 << 57) + (1 << 62);
    let bishops: u64 = (1 << 2) + (1 << 5) + (1 << 58) + (1 << 61);
    let queens: u64 = (1 << 3) + (1 << 59);
    let kings: u64 = (1 << 4) + (1 << 60);

    let white_pieces: u64 = (1 << 16) - 1;
    let black_pieces: u64 = ((1 << 16) - 1) << 48;
    let en_passant: i8 = 0;
    let castling: u8 = (1 << 4) - 1;
    let turn: u8 = WHITE;
    let promoted: i8 = -1;
    let promoted_piece: u8 = 0;
    let current_move: u64 = 0;

    Box::new(Board{
        pawns,
        knights,
        rooks,
        bishops,
        queens,
        kings,

        white_pieces,
        black_pieces,

        en_passant,
        castling,
        turn,
        promoted,
        promoted_piece,
        current_move
    })
}
// Convert immutable &Board to mutable Board
pub fn copy_board(board: &Box<Board>) -> Box<Board> {
    let pawns: u64 = board.pawns;
    let rooks: u64 = board.rooks;
    let knights: u64 = board.knights;
    let bishops: u64 = board.bishops;
    let queens: u64 = board.queens;
    let kings: u64 = board.kings;

    let white_pieces: u64 = board.white_pieces;
    let black_pieces: u64 = board.black_pieces;
    let en_passant: i8 = board.en_passant;
    let castling: u8 = board.castling;
    let turn: u8 = board.turn;
    let promoted: i8 = board.promoted;
    let promoted_piece: u8 = board.promoted_piece;
    let current_move: u64 = board.current_move;

    Box::new(Board {
        pawns,
        rooks,
        knights,
        bishops,
        queens,
        kings,

        white_pieces,
        black_pieces,

        en_passant,
        castling,
        turn,
        promoted,
        promoted_piece,
        current_move
    })
}
